calyx-ir/src/from_ast.rs

```diff
@@ -568,12 +568,160 @@ fn build_static_guard(
 
 ///////////////// Control Construction /////////////////////////
 
+// builds static_seq based on stmts, attributes, and latency
+fn build_static_seq(
+    stmts: Vec<ast::Control>,
+    attributes: Attributes,
+    latency: u64,
+    builder: &mut Builder,
+) -> CalyxResult<StaticControl> {
+    let mut s = StaticControl::seq(
+        stmts
+            .into_iter()
+            .map(|c| build_static_control(c, builder))
+            .collect::<CalyxResult<Vec<_>>>()?,
+        latency,
+    );
+    *s.get_mut_attributes() = attributes;
+    Ok(s)
+}
+
+fn build_static_par(
+    stmts: Vec<ast::Control>,
+    attributes: Attributes,
+    latency: u64,
+    builder: &mut Builder,
+) -> CalyxResult<StaticControl> {
+    let mut p = StaticControl::par(
+        stmts
+            .into_iter()
+            .map(|c| build_static_control(c, builder))
+            .collect::<CalyxResult<Vec<_>>>()?,
+        latency,
+    );
+    *p.get_mut_attributes() = attributes;
+    Ok(p)
+}
+
+fn build_static_if(
+    port: ast::Port,
+    tbranch: ast::Control,
+    fbranch: ast::Control,
+    attributes: Attributes,
+    latency: u64,
+    builder: &mut Builder,
+) -> CalyxResult<StaticControl> {
+    let mut con = StaticControl::if_(
+        ensure_direction(
+            get_port_ref(port, builder.component)?,
+            Direction::Output,
+        )?,
+        Box::new(build_static_control(tbranch, builder)?),
+        Box::new(build_static_control(fbranch, builder)?),
+        latency,
+    );
+    *con.get_mut_attributes() = attributes;
+    Ok(con)
+}
+
+fn build_static_repeat(
+    num_repeats: u64,
+    body: ast::Control,
+    builder: &mut Builder,
+    attributes: Attributes,
+) -> CalyxResult<StaticControl> {
+    let body = build_static_control(body, builder)?;
+    let total_latency = body.get_latency() * num_repeats;
+    let mut scon =
+        StaticControl::repeat(num_repeats, total_latency, Box::new(body));
+    *scon.get_mut_attributes() = attributes;
+    Ok(scon)
+}
+
+// checks whether `control` is static
+fn build_static_control(
+    control: ast::Control,
+    builder: &mut Builder,
+) -> CalyxResult<StaticControl> {
+    let sc = match control {
+        ast::Control::Enable {
+            comp: component,
+            attributes,
+        } => match builder.component.find_group(component) {
+            Some(_) => {
```

Comment:
The `match` can probably be replaced with a conditional for the erroneous path:
```rust
if builder.component.find_group(component).is_some() {
  return Err(...)
}
// Normal case
```

In general avoid nesting using `match` when possible